op-node: Add Metrics

trianglesphere opened the pull request:
**Description**

This adds one new metric, `sequencer_active` to report if the sequencer is active or not.

This also modifies `transactions_sequenced_total`. It is transformed from a Gauge (incorrectly called a counter previously) to a CounterVec. It is labelled with `type` where the value is `l1_info_deposit`, `deposit`, and `txns` to differentiate the different kinds of transactions.

trianglesphere committed:
op-node: Create metrics to record if sequencer is active

trianglesphere committed:
op-node: Update transactions_sequenced_total metric

tynes commented on op-node/metrics/metrics.go:
```diff
@@ -261,12 +268,11 @@ func NewMetrics(procName string) *Metrics {
 			Help:      "Count of total transactions sequenced",
-		}),
-
+		}, []string{"type"}),
```
Will this break a grafana dashboard? Its ok if it does, just need to loop in the right people to update it

trianglesphere replied:
Potentially. It would add a couple new timeseries under the same metric which may cause some updates. I doubt this would break alerts.

protolambda replied:
I think it's fine, I'll update our internal dashboards to handle the change

tynes commented on op-node/metrics/metrics.go:
```diff
@@ -516,8 +530,10 @@ func (m *Metrics) RecordDerivedBatches(batchType string) {
-	m.TransactionsSequencedTotal.Add(float64(count))
+func (m *Metrics) CountSequencedTxsInBlock(txns int, deposits int) {
+	m.TransactionsSequencedTotal.WithLabelValues("l1_info_deposit").Add(1)
+	m.TransactionsSequencedTotal.WithLabelValues("deposits").Add(float64(deposits - 1))
```
Interop is adding another system tx that backruns the L1 info deposits tx. I would prefer to just count all deposits including L1 attributes, so this doesn't ever get out of sync. We don't want hardfork branching logic in here. It also assumes that there is always a deposit tx, which is true and unlikely to change, but this is fragile to that app logic detail given the subtraction

trianglesphere replied:
Fair that this is annoying to do across HF. I do like the idea of splitting out system transactions from user transactions, but even the difference from deposits to normal L2 transactions provides useful data.

tynes commented on op-node/rollup/engine/build_seal.go:
```diff
@@ -106,10 +107,17 @@ func (eq *EngDeriver) onBuildSeal(ev BuildSealEvent) {
 	eq.metrics.RecordSequencerBuildingDiffTime(buildTime - time.Duration(eq.cfg.BlockTime)*time.Second)
 
 	txnCount := len(envelope.ExecutionPayload.Transactions)
```
Its possible that a helper function on the envelope or execution payload could dedupe code, i feel like we do deposit count elsewhere in the codebase, but also it may be on different data structures

trianglesphere replied:
I found one inside this package and used it

tynes commented:
These changes are generally good @trianglesphere, thank you. Left a few comments, curious what you think

trianglesphere commented:
@tynes I cleaned up some of the logic to avoid duplicating code and simplified the deposit count. Ideally we'd be able to split out each type of txn, but having the deposit vs non deposit split already is a big step forward.

protolambda commented:
/ci authorize 414e4fb540c1cb8df030123c83dcf48fb6d10c12

protolambda committed:
Merge branch 'develop' into metrics

protolambda commented:
/ci authorize 4a60f1bb59d79e83bff9f24703876b5074d9aaae
